### ISSUE TITLE
FlaxUNet2DConditionModel very slow to JIT due to huge jaxpr

### ISSUE THREAD

**Issue by SamPruden** [huggingface/diffusers#1874]
Even this trivial example of `FlaxUNet2DConditionModel` with a single block produces 3,319 lines of jaxpr, leading to compile times in the 10s of seconds.

```
model = diffusers.FlaxUNet2DConditionModel(
    sample_size = 32,
    in_channels = 4,
    out_channels = 4,
    down_block_types = ("DownBlock2D"),
    up_block_types = ("UpBlock2D"),
    block_out_channels = (32,),
    layers_per_block = 2,
)

rng = jax.random.PRNGKey(0)
jaxpr = jax.make_jaxpr(model.init)(rng, jnp.zeros((1, 4, 32, 32)), 0.5, None)
print(len(str(jaxpr).splitlines()))
```

The default `down_block_types = ("CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D")` produces 21,047 lines of jaxpr! A CPU compile on the laptop is taking about 4 minutes. 

The JAX FAQ warns that "You can expect slow compilation if the output is many hundreds or thousands of lines long", which this obviously is.

I don't know whether anything can actually be done to reduce this or whether the model is simply inherently that complex, but it might be good to look at optimising this if that hasn't been done already. This isn't an absolute show-stopper, but it does make development pretty inconvenient and annoying.

**Comment by patrickvonplaten** [comment]
Hey @SamPruden,

I actually very much agree here! Note that the Unet is pretty complicated, so I'm not sure how much we can do here, but we've also noticed that it's extremely slow to compile. 

cc @patil-suraj @pcuenca @yiyixuxu here.

Also maybe @cgarciae do you have any ideas by any chance how we could optimize this?

**Comment by SamPruden** [comment]
I had a quick scan of the source for optimisation opportunities. A couple of things stood out, but none that would substantially improve the test cases I measured for this post. They both involve moving loops from python to lax. 

The blocks loop over `layers_per_block`. This could be moved into a lax loop, however the first iteration has a different shape (`in_channels = self.in_channels if i == 0 else self.out_channels`), so would probably need to be lifted out of the loop. That means that this wouldn't benefit the common case of two layers, but it would help if people went above two. The `in_channels == out_channels` case could be detected and taken advantage of, but I expect that's rare.

Similarly, the loop over blocks could be improved in rare cases. If multiple identical blocks (including shapes) exist consecutively they could be merged into lax loops.

The JAX FAQ makes this recommendation:

> Sometimes it isn’t obvious how to rewrite your code to avoid Python loops because your code makes use of many arrays with different shapes. The recommended solution in this case is to make use of functions like [jax.numpy.where()](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.where.html#jax.numpy.where) to do your computation on padded arrays with fixed shape. The JAX team is exploring a “masking” transformation to make such code easier to write.

However, I assume that comes with a runtime performance (and memory!) penalty which would be unacceptable, unless the compiler manages to optimise out the `where`. If this were a reasonable avenue to go down, this could be substantial. It could take a factor two to four out of the `jaxpr` size, although frankly if you take a factor of 4 out it's still very large.

Everything else looked fairly sensible to me, unfortunately. However, you guys have far more expertise in this than I do, so hopefully there's something to find.

**Comment by cgarciae** [comment]
Hey, will discuss this with the Flax team. The main trick to reduce compile times is to use `nn.scan` over layers but it's not easy to apply this to the UNet architecture.

**Comment by cgarciae** [comment]
Hey! Sorry for the delay. I spoke with someone from the Flax team and the proposed solution in this case would be to use `nn.scan` over intermediate components at the same resolution. For example in `FlaxDownBlock2D` we have:

https://github.com/huggingface/diffusers/blob/a43bdd01cd6ebf698b9d1d684e796b80a67843ae/src/diffusers/models/unet_2d_blocks_flax.py#L146-L148

This logic could be abstracted into its own Module and `scan`-ed to both avoid the for loop and reduce compile time.

**Comment by SamPruden** [comment]
> Hey! Sorry for the delay. I spoke with someone from the Flax team and the proposed solution in this case would be to use `nn.scan` over intermediate components at the same resolution. For example in `FlaxDownBlock2D` we have:
> 
> https://github.com/huggingface/diffusers/blob/a43bdd01cd6ebf698b9d1d684e796b80a67843ae/src/diffusers/models/unet_2d_blocks_flax.py#L146-L148
> 
> This logic could be abstracted into its own Module and `scan`-ed to both avoid the for loop and reduce compile time.

I wonder how much that will help in practice. When I took a quick look at it, I concluded that it would probably be quite rare to have multiple consecutive components of the same resolution to scan over. However, maybe I'm making bad assumptions about typical hyperparameters.

That `self.resnets` has `num_layers` elements, and the first layer is (usually) a different shape from the others. The default value for `num_layers` (and the value I was using in the slow tests above) is `2` - so there's no actual gain here. The `scan` would help if that was set above `2`, but I don't know whether that's usual.

I reached similar conclusions about the other loops. Consecutive blocks are usually different shapes, so scanning doesn't appear to be useful for the common case.

There are certainly ways I can see to optimise this in the general case, but I don't know that any help with the common case.

**Comment by cgarciae** [comment]
If `num_layers` is `1` you'd have no gain I've been told, but maybe you'd gain something from it being `2` on above? The main benefit for compilation as I understand it is not only that there are less things to compile but also that scanned code is not inlined with the rest of the program so XLA will have an easier time.

Sadly the default for this layer seems to be `1` (maybe the config object has a different default).

**Comment by SamPruden** [comment]
The block itself defaults `num_layers` to `1`, but in UNet it gets passed `layers_per_block`

https://github.com/huggingface/diffusers/blob/a43bdd01cd6ebf698b9d1d684e796b80a67843ae/src/diffusers/models/unet_2d_condition_flax.py#L163-L167

Which defaults to `2`.

https://github.com/huggingface/diffusers/blob/a43bdd01cd6ebf698b9d1d684e796b80a67843ae/src/diffusers/models/unet_2d_condition_flax.py#L106

So I think we can reasonably say these values default to `2`.

However, I'm not quite sure how we could gain from `scan` in the `2` case. The pattern across all of the blocks looks like this.

https://github.com/huggingface/diffusers/blob/a43bdd01cd6ebf698b9d1d684e796b80a67843ae/src/diffusers/models/unet_2d_blocks_flax.py#L57-L66

My understanding is that if `in_channels` is different for `i == 0`, that can't\* be folded into the scan. In UNet, `in_channels` is usually different from `out_channels`, so the common case is that `num_layers == 2` would look like one special case, followed by a `scan` over `1` item.  `num_layers == 3` would have a `scan` over `2` items and be beneficial.

> The main benefit for compilation as I understand it is not only that there are less things to compile but also that scanned code is not inlined with the rest of the program so XLA will have an easier time.

This is something that I wasn't aware of. If there's somehow a benefit to `scan` even when only looping once, that would be cool!

\* Can't is a strong word. It probably could by using the masking trick that the FAQ [recommends](https://jax.readthedocs.io/en/latest/faq.html#jit-decorated-function-is-very-slow-to-compile). However, I believe that would probably have a small runtime cost, and I doubt we want to trade runtime speed/memory for compilation speed.
